remote-source/main.go

```diff
@@ -0,0 +1,277 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/auto"
+	"github.com/pulumi/pulumi/sdk/v3/go/auto/optdestroy"
+	"github.com/pulumi/pulumi/sdk/v3/go/auto/optrefresh"
+	"github.com/pulumi/pulumi/sdk/v3/go/auto/optup"
+	"gopkg.in/yaml.v3"
+)
+
+type Project struct {
+	Location string `yaml:"location"`
+	Name     string `yaml:"name"`
+	Path     string `yaml:"path,omitempty"`
+	Branch   string `yaml:"branch,omitempty"`
+}
+
+type Environment struct {
+	Region          string  `yaml:"region"`
+	Organization    string  `yaml:"organization"`
+	StackName       string  `yaml:"stackName"`
+	BaseProject     Project `yaml:"baseProject"`
+	PlatformProject Project `yaml:"platformProject"`
+	AppProject      Project `yaml:"appProject"`
+	DataProject     Project `yaml:"dataProject"`
+}
+
+func main() {
+	// Determine mode of operation; default is to refresh/update
+	destroy := false
+	argsWithoutProg := os.Args[1:]
+	if len(argsWithoutProg) > 0 {
+		if argsWithoutProg[0] == "destroy" {
+			destroy = true
+		}
+	}
+
+	// Get configuration data
+	data, err := os.ReadFile("config.yaml")
+	if err != nil {
+		fmt.Printf("Error reading configuration file: %v\n", err)
+		os.Exit(1)
+	}
+	var env Environment
+	err = yaml.Unmarshal(data, &env)
+	if err != nil {
+		fmt.Printf("Error parsing configuration information: %v\n", err)
+		os.Exit(1)
+	}
+
+	// Set up context
+	ctx := context.Background()
+
+	// Call the setupStack function to set up each of the four stacks
+	// Set up base stack
+	baseStack, err := setupStack(ctx, env.Organization, env.StackName, env.BaseProject)
+	if err != nil {
+		fmt.Printf("Error encountered setting up %s stack: %v\n", env.BaseProject.Name, err)
+		os.Exit(1)
+	}
```

Comment:
It would be lovely to avoid the repeated error handling. I would try to do something similar for the "refresh + update" and "destroy" sections.
```suggestion
	// Call the setupStack function to set up each of the four stacks
	setup := func(p Project) auto.Stack {
		stack, err := setupStack(ctx, env.Organization, env.StackName, project)
		if err != nil {
			fmt.Printf("Error encountered setting up %s stack: %v\n",
				p.Name, err)
			os.Exit(1)
		}
		return stack
	}
	
	baseStack := setup(env.BaseProject)
	platformStack := setup(env.PlatformProject)
	dataStack := setup(env.DataProject)
	appStack := setup(env.AppProject)
```

Since this is example code, we can also move the "print error + exit" to their respective functions, but I like the `lambda` approach better. That way we only exit from `main`.